### PR TITLE
Fix: don't allow free-wagon-chains to exceed max-train-length

### LINKED ISSUE #8295: Game crashes trying to sell huge wagon chain
## Version of OpenTTD
1.10.3 Windows 10 64-bit
## Expected result
The game does not crash.
## Actual result
The game crashes after roughly two months in-game. The exact date does not matter - you can wait for a monthly autosave, load it, and the game will crash in two months.
## Steps to reproduce
1. Load the save.
2. Wait for two months to pass.
3. Get the crash.

Granted I have NewGRFs installed, but the game shouldn't crash like that. In debug mode, there is no relevant errors prior to crash.

[Savegame](https://github.com/OpenTTD/OpenTTD/files/5065970/autosave0.zip)

### PR DESCRIPTION
**Pull request by TrueBrain**
This doesn't fix  #8295, but prevents any future AIs from ever getting in this state.

## Motivation / Problem

In #8295 the AI created a very very very long free-wagon chain. This made me wonder why this was allowed. Trains are limited by a max length after all, so why not free-wagon chains?

This PR limits free-wagon chains under the same conditions trains are limited. 

## Description

This makes no sense, that a free-wagon-chain could be larger than
the maximum length of a train, as you cannot put an engine in
front of that anyway. And it prevents run-away AIs making very
silly long free-wagon-chains.

## Limitations

- Ideally, it would stop just before the train limit, so you can always put an engine in front of it. But as this limit is based on pixel-length, and engines can have all sorts of lengths, I made the choice not to do that.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

File: src/train_cmd.cpp [hunk 1]
```diff
@@ -646,8 +646,9 @@ static CommandCost CmdBuildRailWagon(TileIndex tile, DoCommandFlag flags, const
 					w->engine_type == e->index &&   ///< Same type
 					w->First() != v &&              ///< Don't connect to ourself
 					!(w->vehstatus & VS_CRASHED)) { ///< Not crashed/flooded
-				DoCommand(0, v->index | 1 << 20, w->Last()->index, DC_EXEC, CMD_MOVE_RAIL_VEHICLE);
-				break;
+				if (!DoCommand(0, v->index | 1 << 20, w->Last()->index, DC_EXEC, CMD_MOVE_RAIL_VEHICLE).Failed()) {
```
Comment (J0anJosep) [inline src/train_cmd.cpp]: Succeeded instead of no failed would bé cleaner.

File: src/train_cmd.cpp [hunk 2]
```diff
@@ -985,7 +986,7 @@ static CommandCost CheckTrainAttachment(Train *t)
 		t->SetNext(nullptr);
 
 		/* Don't check callback for articulated or rear dual headed parts */
-		if (!t->IsArticulatedPart() && !t->IsRearDualheaded()) {
+		if (t->IsEngine() && !t->IsArticulatedPart() && !t->IsRearDualheaded()) {
```
Comment (frosch123) [inline src/train_cmd.cpp]: This is supposed to check whether this is a real train. I.e. IsEngine should check the first vehicle.
However, "t" can be about anything here, and you do not want to call the attach-callback for free wagon chains.

File: src/train_cmd.cpp [hunk 3]
```diff
@@ -959,12 +960,25 @@ static CommandCost CheckNewTrain(Train *original_dst, Train *dst, Train *origina
...
-	if (t == nullptr || t->Next() == nullptr || !t->IsEngine()) return CommandCost();
+	if (t == nullptr || t->Next() == nullptr) return CommandCost();
 
 	/* The maximum length for a train. For each part we decrease this by one
 	 * and if the result is negative the train is simply too long. */
 	int allowed_len = _settings_game.vehicle.max_train_length * TILE_SIZE - t->gcache.cached_veh_length;
 
+	/* For free-wagon chains, check if they are within the max_train_length limit. */
+	if (!t->IsEngine()) {
+		t = t->Next();
```
Comment (TrueBrain) [inline src/train_cmd.cpp:971]: I do not know if I should call `head->InvalidateNewGRFCache();` before doing this, honestly. I hope someone can help me out there.

In fact, I am not even sure if I should also do the `SetNext()` stuff as done lower. Normally this function wasn't executed for free-wagon-chains, so I assume the `Next()` chain is valid, but I have honestly no idea.

It works in-game, so we have that at least working for us ..

Comment (frosch123) [reply src/train_cmd.cpp:971]: All good.
* InvalidateNewGRFCache() is potato/potato here. It's only needed for the callbacks below.
* SetNext() is only changed temporary during the callback below, it's immediately restored.

**Review by frosch123 — APPROVED** [review]

**Pull request merged**